units/github_status_api.go

```diff
@@ -0,0 +1,320 @@
+	}
+	if respStatus == nil {
+		j.AddError(errors.New("nil response from github"))
+	}
+}
+
+func (j *githubStatusUpdateJob) fetch(status *githubStatus) {
+	patchVersion := j.FetchID
+	if j.UpdateType == "build" {
+		b, err := build.FindOne(build.ById(j.FetchID))
+		if err != nil {
+			j.AddError(err)
+		}
+		if b == nil {
+			j.AddError(errors.New("can't find build"))
+		}
+		if j.HasErrors() {
+			return
+		}
+
+		patchVersion = b.Version
+		status.context = fmt.Sprintf("evergreen-%s", b.BuildVariant)
+		status.description = taskStatusToDesc(b)
+		status.urlPath = fmt.Sprintf("/build/%s", b.Id)
+
+		switch b.Status {
+		case evergreen.BuildSucceeded:
+			status.ghStatus = githubStatusSuccess
+
+		case evergreen.BuildFailed:
+			status.ghStatus = githubStatusFailure
+
+		default:
+			j.AddError(errors.New("build status is pending; refusing to update status"))
+			return
+		}
+	}
+
+	patchDoc, err := patch.FindOne(patch.ByVersion(patchVersion))
+	if err != nil {
+		j.AddError(err)
+	}
+	if patchDoc == nil {
+		j.AddError(errors.New("can't find patch"))
+	}
+	if j.HasErrors() {
+		return
+	}
+
+	if j.UpdateType == "patch" {
+		status.urlPath = fmt.Sprintf("/version/%s", patchVersion)
+		status.context = "evergreen"
+
+		switch patchDoc.Status {
+		case evergreen.PatchSucceeded:
+			status.ghStatus = githubStatusSuccess
+			status.description = fmt.Sprintf("finished in %s", patchDoc.FinishTime.Sub(patchDoc.StartTime).String())
+
+		case evergreen.PatchFailed:
+			status.ghStatus = githubStatusFailure
+			status.description = fmt.Sprintf("finished in %s", patchDoc.FinishTime.Sub(patchDoc.StartTime).String())
+
+		case evergreen.PatchCreated:
+			status.ghStatus = githubStatusPending
+			status.description = "preparing to run tasks"
+
+		case evergreen.PatchStarted:
+			status.ghStatus = githubStatusPending
+			status.description = "tasks are running"
+
+		default:
+			j.AddError(errors.New("unknown patch status"))
+			return
+		}
+	}
+
+	status.owner = patchDoc.GithubPatchData.BaseOwner
+	status.repo = patchDoc.GithubPatchData.BaseRepo
+	status.prNumber = patchDoc.GithubPatchData.PRNumber
+	status.ref = patchDoc.GithubPatchData.HeadHash
+}
+
+func (j *githubStatusUpdateJob) Run() {
+	defer j.MarkComplete()
+
+	adminSettings, err := admin.GetSettings()
+	if err != nil {
+		j.AddError(errors.Wrap(err, "error retrieving admin settings"))
+	}
+	if adminSettings.ServiceFlags.GithubPRTestingDisabled {
+		grip.InfoWhen(sometimes.Percent(evergreen.DegradedLoggingPercent), message.Fields{
+			"job":     githubStatusUpdateJobName,
+			"message": "github pr testing is disabled, not updating status",
+		})
+		j.AddError(errors.New("github pr testing is disabled, not updating status"))
+		return
+	}
+
+	status := githubStatus{}
+	j.fetch(&status)
+	if j.HasErrors() {
+		return
+	}
+
+	j.sendStatusUpdate(&status)
+	if j.HasErrors() {
+		grip.Alert(message.Fields{
+			"message": "github API failure",
+			"job":     j.ID(),
```

Comment:
I worry that it'll be hard to backtrack this to an actual PR and/or patch version

Reply:
The status object contains the pr number, repository, and git ref. That should be more than enough. Also adding fetch id and fetch type

Reply:
My preference is to avoid needing to parse out the string ourselves, and just put that in the fields?